python/pdo/contract/replication.py

```diff
@@ -46,68 +43,78 @@
 __stop_service__ = False
 
 # -----------------------------------------------------------------
-def start_replication_service():
-
-    # start the manager
-    __replication_manager_executor__.submit(__replication_manager__)
+def start_replication_service(config = {}) :
+    # make sure the service is only started once
+    global replication_task_queue
+    with replication_thread_lock :
+        if replication_task_queue :
+            return
+        else :
+            replication_task_queue = queue.Queue()
+
+    global number_of_service_threads
+    number_of_service_threads = config.get("replication_worker_threads", number_of_service_threads)
+
+    logger.debug('start replication manager threads')
+    for i in range(config.get("replication_service_threads", 1)) :
+        thread = threading.Thread(target=__replication_manager__)
+        thread.daemon = True
+        thread.start()
+
+        with replication_thread_lock :
+            replication_threads.append(thread)
 
 # -----------------------------------------------------------------
-def stop_replication_service():
+def stop_replication_service() :
+    logger.debug("stop replication service")
 
     global __stop_service__
     __stop_service__ = True
 
-    #shutdown replication executor
-    __replication_manager_executor__.shutdown(wait=True)
+    for thread in replication_threads[:] :
+        thread.join(timeout=5.0)
 
 # -----------------------------------------------------------------
 def add_replication_task(task):
-
-    __pending_replication_tasks_manager_queue__.put(task)
+    replication_task_queue.put(task)
```

Comment:
Shouldn't this write be done with the `replication_task_lock`?

Reply:
afaik, python queues are thread safe. (The [Queue](https://docs.python.org/3/library/queue.html#queue.Queue) class in this module implements all the required locking semantics.)

Reply:
Ah, thanks! Learned something new :)